tests/test_charm.py

```diff
@@ -105,3 +109,107 @@ def test_new_unit_has_password(self):
             config["MYSQL_ROOT_PASSWORD"],
             self.harness.charm._stored.mysql_setup["MYSQL_ROOT_PASSWORD"],
         )
+
+    def test_provides(self):
+        with patch("charm.MySQLCharm.mysql") as mock_version:
+            VERSION = "mysql 8.0.22"
+            mock_version.return_value = VERSION
+            MySQLCharm.mysql.version = mock_version
+
+            info = {
+                "app_name": "mysql",
+                "host": "mysql-0.mysql-endpoints",
+                "port": 3306,
+                "user_name": "root",
+                "mysql_root_password": "D10S!",
+            }
+            db_info = Mock(return_value=info)
+            MySQLCharm.db_info = db_info
+
+            self.assertTrue(isinstance(self.harness.charm.provides, dict))
+            self.assertTrue("provides" in self.harness.charm.provides)
+            self.assertTrue(
+                isinstance(self.harness.charm.provides["provides"], dict)
+            )
+            self.assertEqual(
+                self.harness.charm.provides["provides"]["mysql"], VERSION
+            )
+            self.assertTrue(
+                isinstance(self.harness.charm.provides["config"], dict)
+            )
+            self.assertEqual(
+                self.harness.charm.provides["config"]["app_name"], "mysql"
+            )
+            self.assertEqual(
+                self.harness.charm.provides["config"]["host"],
+                "mysql-0.mysql-endpoints",
+            )
+            self.assertEqual(
+                self.harness.charm.provides["config"]["port"], 3306
+            )
+            self.assertEqual(
+                self.harness.charm.provides["config"]["user_name"], "root"
+            )
+            self.assertEqual(
+                self.harness.charm.provides["config"]["mysql_root_password"],
+                "D10S!",
+            )
+
+    def test__on_start(self):
+        self.harness.set_leader(False)
+        self.assertEqual(self.harness.charm.on.start.emit(), None)
+
+        with patch("mysqlserver.MySQL.is_ready") as mock_is_ready:
+            mock_is_ready.return_value = False
+            self.harness.set_leader(True)
+            config = {
+                "MYSQL_ROOT_PASSWORD": "D10S!",
+            }
+            self.harness.update_config(config)
+            self.assertEqual(self.harness.charm.on.start.emit(), None)
+
+        with patch("mysqlserver.MySQL.is_ready") as mock_is_ready:
+            self.harness.set_leader(True)
+            mock_is_ready.return_value = True
+            config = {
+                "MYSQL_ROOT_PASSWORD": "D10S!",
+            }
+            self.harness.update_config(config)
+            self.harness.charm.on.start.emit()
+            self.assertEqual(
+                type(self.harness.charm.unit.status), ActiveStatus
+            )
+
+    def test__update_status(self):
```

Comment:
Another stylistic suggestion. Seems to me there are really three different test cases in this one test.
- Test non leader unit always sets active status with no message
- Test leader unit sets some message on failed status
- Test leader unit sets some other message on success status.
By splitting the tests into three separate tests has IMHO the following benefits
- Tests are smaller
- Tests names can be chosen to be very descriptive
- Failure of a test immediately indicates what is failing because the tests are specific and the names are descriptive.

Reply:
I agree with you.
done